File: src/my_ude_py/scripts/navigation_server.py
#!/usr/bin/env python3
import rclpy
import time
import math
from rclpy.node import Node
from geometry_msgs.msg import Point
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.executors import MultiThreadedExecutor

from ude_interface.action import Navigate

DISTANCE_THRESHOLD = 0.125

class NavigationServerNode(Node): 
    def __init__(self):
        super().__init__("navigation_server") 
        self.subscriber_  = self.create_subscription(Point, "robot_position", self.update_robot_position, 1)
        self.robot_current_position = None 
        self.action_server_= ActionServer(self, Navigate, 'navigate', 
                                        execute_callback=self.execute_callback,
                                        goal_callback=self.goal_callback,
                                        cancel_callback=self.cancel_callback)
        self.get_logger().info("Navigation Server Node has been started")

    def update_robot_position(self, point):
        self.robot_current_position = [point.x, point.y, point.z]

    def handle_accepted_callback(self, goal_handle):
        """ Maneja el goal aceptado """
        goal_id = goal_handle.goal_id
        self._logger.info("Goal request with ID: " + str(goal_handle.goal_id) + " has been accepted.")

    def goal_callback(self, goal_request):
        """ Maneja el goal enviado por el cliente """
        self.get_logger().info("Received goal request with ID: " + str(goal_request))
        x = goal_request.goal_point.x
        y = goal_request.goal_point.y
        z = goal_request.goal_point.z       
        return GoalResponse.ACCEPT

    def cancel_callback(self, goal_handle):
        """Accept or reject a client request to cancel an action."""
        self.get_logger().info("Received cancel request for goal request with ID: " + str(goal_handle.goal_id))
        return CancelResponse.ACCEPT

    
    def execute_callback(self, goal_handle):
        self.get_logger().info("Goal request with ID: " + str(goal_handle.goal_id) + " has been accepted.")    
        self.get_logger().info("Executing goal")    
        start_time = self.get_clock().now().to_msg().sec 
        feedback_msg = Navigate.Feedback()
        result = Navigate.Result()
        robot_goal_point = [goal_handle.request.goal_point.x, 
							goal_handle.request.goal_point.y,
							goal_handle.request.goal_point.z]
        
        
        self.get_logger().info("Goal Point: " + str(robot_goal_point))

        while self.robot_current_position == None:
            self.get_logger().warn("Robot Point Not Detected")
            time.sleep(3.0)

        distance_to_goal = math.dist(self.robot_current_position, robot_goal_point)
        feedback_msg = Navigate.Feedback()
        
        if(self.robot_current_position==robot_goal_point):
            self.get_logger().info("Goal aborted")
            goal_handle.abort()
            result.elapsed_time = float(self.get_clock().now().to_msg().sec - start_time)
            return result
        
        while(distance_to_goal > DISTANCE_THRESHOLD):
            if goal_handle.is_cancel_requested:
                self.get_logger().info("Goal canceled")
                result.elapsed_time = float(self.get_clock().now().to_msg().sec - start_time)
                goal_handle.canceled()
                return result
            distance_to_goal = math.dist(self.robot_current_position, robot_goal_point)
            feedback_msg.distance_to_point = distance_to_goal
            goal_handle.publish_feedback(feedback_msg)
            self.get_logger().info("Publish feedback")
            time.sleep(1.0)

        self.get_logger().info("Goal succeeded")
        goal_handle.succeed()
        result.elapsed_time = float(self.get_clock().now().to_msg().sec - start_time)
        return result

def main(args=None):
    rclpy.init(args=args)
    node = NavigationServerNode() 
    executor = MultiThreadedExecutor()
    rclpy.spin(node, executor=executor)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_ude_cpp/src/navigation_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

#include "ude_interface/action/navigate.hpp"
#include "geometry_msgs/msg/point.hpp"

using std::placeholders::_1;
using std::placeholders::_2;
using geometry_msgs::msg::Point;

typedef ude_interface::action::Navigate Navigate;
typedef rclcpp_action::ServerGoalHandle<Navigate> GoalHandle;

const float DIST_THRESHOLD = 0.125; 

class NavigationServerNode : public rclcpp::Node
{
public:
    NavigationServerNode() : Node("navigation_server") 
    {
        robot_position_ = Point();
        subscription_ = this->create_subscription<Point>("robot_position", 10, 
                std::bind(&NavigationServerNode::update_robot_position, this, std::placeholders::_1));
        action_server_= rclcpp_action::create_server<Navigate>(this ,"navigate",
                        std::bind(&NavigationServerNode::handle_goal, this ,  _1, _2),
                        std::bind(&NavigationServerNode::handle_cancel, this ,  _1),
                        std::bind(&NavigationServerNode::handle_accepted, this ,  _1));
    	RCLCPP_INFO(this->get_logger(), "Navigation Server Node has been started.");
    }

private:
    void update_robot_position(const Point & msg)
    {
      robot_position_ = msg;
    }

    rclcpp_action::GoalResponse handle_goal(const rclcpp_action::GoalUUID &uuid,
                                            std::shared_ptr<const Navigate::Goal> goal)
    {   
        (void)uuid;
        (void)goal;
        //RCLCPP_INFO(get_logger(), "Received goal request with ID: %i", uuid);
        RCLCPP_INFO(get_logger(), "Received goal request with ID");
        /*float x = goal->goal_point.x;
        float y = goal->goal_point.y;
        float z = goal->goal_point.z;*/
        rclcpp_action::GoalResponse result = rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
        return result;
    }

    rclcpp_action::CancelResponse handle_cancel(const std::shared_ptr<rclcpp_action::ServerGoalHandle<Navigate>> goal_handle)
    {
        (void)goal_handle;
        //RCLCPP_INFO(get_logger(), "Received cancel request for goal request with ID: %i", goal_handle->get_goal_id());
        RCLCPP_INFO(get_logger(), "Received cancel request for goal request with ID");
        rclcpp_action::CancelResponse result = rclcpp_action::CancelResponse::ACCEPT;
        return result;
    }

    void handle_accepted(const std::shared_ptr<rclcpp_action::ServerGoalHandle<Navigate>> goal_handle)
    {
        //RCLCPP_INFO(get_logger(), "Goal request with ID %i has been accepted.", goal_handle->get_goal_id());
        RCLCPP_INFO(get_logger(), "Goal request with ID has been accepted.");
        // this needs to return quickly to avoid blocking the executor, so spin up a new thread
        std::thread{std::bind(&NavigationServerNode::execute, this, std::placeholders::_1), goal_handle}.detach();
    }

    void execute(const std::shared_ptr<GoalHandle> goal_handle)
    {
        RCLCPP_INFO(this->get_logger(), "Executing goal");
        auto start_time = rclcpp::Clock().now();
        const auto goal = goal_handle->get_goal();
        auto feedback = std::make_shared<Navigate::Feedback>();
        auto result = std::make_shared<Navigate::Result>();
        feedback->distance_to_point = std::sqrt(
          std::pow(this->robot_position_.x - goal->goal_point.x, 2) + 
          std::pow(this->robot_position_.y - goal->goal_point.y, 2) + 
          std::pow(this->robot_position_.z - goal->goal_point.z, 2));
        auto robot_goal_point = Point();
        robot_goal_point = goal->goal_point;
        rclcpp::Rate loop_rate(1.0);
       
        if(this->robot_position_==robot_goal_point){
            RCLCPP_INFO(this->get_logger(), "Goal aborted");
            result->elapsed_time = (rclcpp::Clock().now() - start_time).seconds();
            goal_handle->abort(result);
            return;
        }


        while(feedback->distance_to_point > DIST_THRESHOLD){
            // Check if there is a cancel request
            if (goal_handle->is_canceling()) {
                result->elapsed_time = (rclcpp::Clock().now() - start_time).seconds();
                goal_handle->canceled(result);
                RCLCPP_INFO(this->get_logger(), "Goal canceled");
                return;
            }
            feedback->distance_to_point = std::sqrt(
            std::pow(this->robot_position_.x - goal->goal_point.x, 2) + 
            std::pow(this->robot_position_.y - goal->goal_point.y, 2) + 
            std::pow(this->robot_position_.z - goal->goal_point.z, 2));

            goal_handle->publish_feedback(feedback);
            RCLCPP_INFO(this->get_logger(), "Publish feedback");
            loop_rate.sleep();
        }
        RCLCPP_INFO(this->get_logger(), "Goal succeeded");
        result->elapsed_time = (rclcpp::Clock().now() - start_time).seconds();
        goal_handle->succeed(result);
        
    }
    Point robot_position_;
    rclcpp_action::Server<Navigate>::SharedPtr action_server_;
    rclcpp::Subscription<Point>::SharedPtr subscription_;

};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NavigationServerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}